### PR TITLE
Add ENS Name routing support for V2 projects

### PR DESCRIPTION
**Pull request by aeolianeth**
## What does this PR do and why?

What it says on the tin.

Closes https://github.com/jbx-protocol/juice-interface/issues/855

## Screenshots or screen recordings

This ENS name has a text record `juicebox_v2` set to `2` (project ID `2`):

https://app.ens.domains/name/chasemcdude.eth/details

| project ID not found | ens fails (i.e. has no text record) | ens resolves |
| --- | --- | ---|
| Screen Shot 2022-04-24 at 12 34 38 pm | Screen Shot 2022-04-24 at 12 45 43 pm | Screen Shot 2022-04-24 at 12 34 52 pm |

## Acceptance checklist

- [ ] I have evaluated the [Approval Guidelines](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#approval-guidelines) for this PR.
- [ ] I have tested this PR in [all supported browsers](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#supported-browsers).

### CODE REVIEW

**Comment by aeolianeth** [comment]
Hey @peripheralist - mind taking a quick look at this if you get a chance?

File: src/Router.tsx [hunk 1]
```diff
@@ -39,6 +41,16 @@ export default function Router() {
         <Route path="/projects">
           <Projects />
         </Route>
+
+        <Route path="/p/:ensName(*.eth)">
```
Comment (aeolianeth) [inline src/Router.tsx:45]: this regex is very permissive and potentially not ideal, but i think we can rely on the Ethers ENS resolver to fail if dodgy input is here.

File: src/components/v2/V2Dashboard/index.tsx [hunk 2]
```diff
@@ -39,19 +38,7 @@ import {
   RESERVED_TOKEN_SPLIT_GROUP,
 } from 'constants/v2/splits'
 
-const parseProjectIdParameter = (projectIdParameter?: string) => {
```
Comment (aeolianeth) [inline src/components/v2/V2Dashboard/index.tsx:42]: Delegate this logic to the Gateway component

File: src/hooks/ProjectENSNameResolver.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import { useEffect, useState } from 'react'
+
+import { BigNumber } from '@ethersproject/bignumber'
+
+import { readProvider } from 'constants/readProvider'
+
+const JUICEBOX_V2_TXT_RECORD_KEY = 'juicebox_v2'
```
Comment (aeolianeth) [inline src/hooks/ProjectENSNameResolver.ts:7]: We'll need some docs (or potentially UI instruction) for setting this TXT record in ENS

**Comment by peripheralist** [comment]
@aeolianeth fun pr! This is huge. Love the structure here, especially the gateway pattern. Nice and clean. Left a few comments and will look at this again tomorrow with fresh eyes

File: src/hooks/ProjectENSNameResolver.ts (hunk 3 above)
Comment (johnnyd-eth) [reply src/hooks/ProjectENSNameResolver.ts:7]: @aeolianeth will it be possible to do programatically?

**Comment by johnnyd-eth** [comment]
Fuck yeah @aeolianeth this is so sick. One non-blocking question from me but otherwise looks clean and rad AF!

**Review by johnnyd-eth — APPROVED** [review]

File: src/hooks/ProjectENSNameResolver.ts (hunk 3 above)
Comment (aeolianeth) [reply src/hooks/ProjectENSNameResolver.ts:7]: yeah i think it actually might be. we might be able to queue up the TX to set the text record on our end. would be dope

Comment (johnnyd-eth) [reply src/hooks/ProjectENSNameResolver.ts:7]: Indeed. Even a modal to buy an ENS name if they don't have one yet 👀

**Comment by aeolianeth** [comment]
Thanks @johnnyd-eth @peripheralist! ❤️ 

@peripheralist I can't see any of your comments - may be waiting for submission in the review draft 😎

**Comment by peripheralist** [comment]
> Thanks @johnnyd-eth @peripheralist! ❤️ 
> 
> 
> 
> @peripheralist I can't see any of your comments - may be waiting for submission in the review draft 😎  

That was exactly why 💀

File: src/hooks/ProjectENSNameResolver.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  ensName,
+}: {
+  ensName?: string
+}) {
+  const [loading, setLoading] = useState<boolean>(false)
+  const [projectId, setProjectId] = useState<BigNumber | null>(null)
+
+  useEffect(() => {
+    const resolve = async () => {
+      setLoading(true)
```
[the remark this answers is not in the document]
Comment (aeolianeth) [reply src/hooks/ProjectENSNameResolver.ts]: gooooood 🙏

File: src/components/v2/V2Dashboard/Gateway.tsx [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import { BigNumber } from '@ethersproject/bignumber'
+import useProjectENSNameResolver from 'hooks/ProjectENSNameResolver'
+import { useParams } from 'react-router-dom'
+import Loading from 'components/shared/Loading'
+
+import V2Dashboard from '.'
+import Dashboard404 from './Dashboard404'
+
+const parseProjectIdParameter = (projectIdParameter?: string) => {
```
[the remark this answers is not in the document]
Comment (aeolianeth) [reply src/components/v2/V2Dashboard/Gateway.tsx:9]: Good question - this function is actually for passing the URL query parameter project ID, which is a string.

ProjectENSNameResolver is already returning a BigNumber: https://github.com/jbx-protocol/juice-interface/blob/e79f6e62bc67d200313c9984d86f2797446132ab/src/hooks/ProjectENSNameResolver.ts#L24

File: src/hooks/ProjectENSNameResolver.ts (hunk 3 above)
Comment (aeolianeth) [reply src/hooks/ProjectENSNameResolver.ts]: > Also kind of a side note, could we just use "juicebox" for the record name? There's a UX element here for anyone who chooses to update the record manually, and adding the _v2 may seem a little awkward considering we don't use this for v1 (or if you don't even know about jb versions)

yeah good call, I was thinking that somehow we could extend this to support V1 and V2. Probably out of scope though.

Renamed it to `juicebox` 👍

File: src/components/v2/V2Dashboard/Gateway.tsx [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  const {
+    projectId: projectIdParameter,
+    ensName,
+  }: { projectId?: string; ensName?: string } = useParams()
+
+  const { loading, projectId: projectIdForName } = useProjectENSNameResolver({
+    ensName,
+  })
+
+  if (loading) return <Loading />
```
[the remark this answers is not in the document]
Comment (aeolianeth) [reply src/components/v2/V2Dashboard/Gateway.tsx:27]: It'll be loading if the ENS name is being resolved - but yep, your comment above https://github.com/jbx-protocol/juice-interface/pull/859#discussion_r857391176 is def an improvement - no loading state if no ENS name (if the route doesn't have a match for `*.eth`)

**Comment by aeolianeth** [comment]
@peripheralist awesome comments thank you - addressed 🏓

**Pull request merged**